### DISCUSSION TITLE
Memory usage in Contact modules

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#16773] [Q&A Modules: General] [answered]
Quick question, Ive got a 'toy' problem which resembles our children's favourite Danish building blocks. There are two blocks next to each other, along the long face as seen below.

![Screenshot from 2021-01-21 09-35-14](https://user-images.githubusercontent.com/2439754/105332003-062cfe00-5bcc-11eb-9c56-d4602e5a79ed.png)

In total there are 1583 nodes on each surface (4928 edges and 3346 triangles if that matters - dont think it does) and trying to get this problem setup and running on my desktop, requires more than 128 Gb of memory ?!?! That seems excessive to me, any suggestions here would be appreciated.

**Comment by makeclean** [comment, marked as answer]
So memory usage is around 18.2 Gb until computing initial stateful property values, and then it rockets up fairly quickly to 128 Gb and then is killed.

**Reply by aeslaughter** [reply]
What type of properties are you using that are stateful? Saving the state requires that values of the material properties are stored for every quadrature point in your mesh. It is common to have a large memory demand with stateful properties.

**Reply by makeclean** [reply]
nothing of note which makes it weird

**Reply by lindsayad** [reply]
With your `dock_slippy.i` input I am not seeing memory usage above about 300 MB

**Reply by lindsayad** [reply]
Replying on this thread so I can leave the other one to talk about convergence. Congrats @makeclean you've unearthed two large memory issues. Here is this ticket: https://github.com/idaholab/moose/issues/16836

**Reply by lindsayad** [reply]
Ok after #16845, for your lego input, I go from this graph:
![fixed-temp](https://user-images.githubusercontent.com/10248304/106531103-6eb88c80-64a2-11eb-9bc8-2f6a94f157c0.png)
which was taken from the middle of initing stateful props, and shortly before I canceled the simulation so my computer wouldn't run out of memory to this graph:
![post-stateful-fixes](https://user-images.githubusercontent.com/10248304/106531166-8bed5b00-64a2-11eb-9e0e-e304a340e007.png)
which was taken *after* initing stateful properties...because now we can get to that point without running out of memory. So 84 GB for stateful properties (and that number was still growing) to 4GB. I'm a big fan. Thanks for your helpful inputs @makeclean !

**Comment by makeclean** [comment]
Some progress, it seems the two objects in contact *must* be separate mesh blocks, is that true MOOSE folks?

**Reply by makeclean** [reply]
If two surfaces were part of a contact group, and those surfaces also had nodes that were part of a fixed 0 displacement boundary, what would happen?

**Reply by aeslaughter** [reply]
@lindsayad Can you answer this contact question?

**Reply by makeclean** [reply]
Ill try and formulate a minimum viable problem that shows the issue.

**Reply by lindsayad** [reply]
What contact formulation are you using? With respect to contact vs. dirichlet condition: the dirichlet condition would win because it runs after constraints. A MWE would be great yea, thanks!

**Reply by makeclean** [reply]
Ok, I can't reproduce the problem with my simple block example, where should I put the mesh from the example above?

**Reply by makeclean** [reply]
Its regular contact not dirichlet

**Reply by lindsayad** [reply]
> If two surfaces were part of a contact group, and those surfaces also had nodes that were part of a fixed 0 displacement boundary, what would happen?

Right here you asked what would happen if you had a surface that was simultaneously designated for contact and was part of a fixed 0 displacement boundary, which is presumably a Dirichlet BC.

**Reply by makeclean** [reply]
oh i see, sorry, bit slow tonight :)

**Comment by makeclean** [comment]
Also, further weirdness, I have what I'm calling a 'docking' problem - but basically I have a 'rubber' hollow sphere (as shown in the image) and a stainless steel 'sphere' I am forcing the displacement as a function of time so that they start seperate and then get into contact later. In this case gravity is acting along the +ve z axis, there is a 0 displacement condition on the flat bottom of the rubber hollow sphere. I also constrain the x,y movement of the sphere with 0 displacement condition, such that the sphere only moves along the z axis. There are 3 contact surfaces; the outside part of the stainless sphere (a), the entry part into the hollow sphere (b) and the inside of the hollow sphere (c) and I have frictional contacts setup for ac and bc. All is great, contact is made the rubber deforms, but eventually we get a convergence failure at the minimum timestep. 

![Screenshot from 2021-01-26 20-58-02](https://user-images.githubusercontent.com/2439754/105904582-3ab01800-6019-11eb-9909-b0b74b808ade.png)

I've tried turning on AMR, but this leads to segfaults and also some weird deformations that you'll see in the next image, and some oddly connected mesh elements. 

![Screenshot from 2021-01-26 21-01-18](https://user-images.githubusercontent.com/2439754/105904890-a8f4da80-6019-11eb-8b5b-868e3ab783a6.png)

So my questions are manifold, but 

1) should contact and AMR work concurrently?
2) Ive tried taking both to second order but this leads to both PETSC segfault-ing
3) should I try just a finer mesh?

Happy to share inputs, can I attach them here?

**Reply by lindsayad** [reply]
> Happy to share inputs, can I attach them here?

If they are MOOSE-only

**Reply by makeclean** [reply]
no, exodus meshes from Cubit im afraid

**Reply by lindsayad** [reply]
Oh I just meant MOOSE-only objects in the input file. How big are the meshes?

**Reply by makeclean** [reply]
Not huge, only 50k ish elements - runs in around 30 mins to failure

**Reply by makeclean** [reply]
on one core

**Reply by lindsayad** [reply]
If they are not ginormous, you can send them to alexander.lindsay@inl.gov. (I don't think you can attach them here on github...if you can, then just do that)

**Reply by lindsayad** [reply]
The dock problem should also show the memory usage problem right? I  see that you are using finite strain

**Reply by makeclean** [reply]
Dock slippy has no memory problems and runs fine. Only the original lego problem had that issue, which I can't reproduce in a small problem. I'll forward on the original file, its not that huge either, hence the original question re: memory usage. The issue with dock_slippy.i is eventually a convergence fail, and the fact that going to 2nd order is an immediate segfault.

**Reply by lindsayad** [reply]
Ok yes I would like to have the really bad memory input. FWIW even dock-slippy, though it only resulted in 300 MB of peak memory usage, showed us a potential bug that I am investigating a fix for right now. Once I am done investigating the memory usage, I will think about the other issues.

**Reply by makeclean** [reply]
Oh is that bug the one about the stress in the moving part comes and goes away?

**Reply by lindsayad** [reply]
No...it's a memory bug. See fix in #16831

**Reply by lindsayad** [reply]
@makeclean with respect to convergence failure for the "normal" simulation (e.g. first order and no AMR), I am not really surprised. Contact is notoriously difficult for Newton because of it's discontinuous function nature. It looks to me like you are using `formulation = kinematic`. I believe that our application developers have found that for frictional models, `tangential_penalty` is the most robust. Perhaps @bwspenc or @recuero can weigh-in on this. I'm also going to attach your input here in case something jumps out to them. (Although they may just turn around and ask for the mesh, which I can forward)
[dock_slippy.txt](https://github.com/idaholab/moose/files/5890432/dock_slippy.txt)

With respect to second order, I believe our mechanics folks run plenty of simulations with second order Lagrange. Again @bwspenc and @recuero can correct me if I'm wrong, but segfaults out of PETSc would be a bit surprising. I can run in debug mode hopefully and identify the exact cause of the segfault.

**Reply by recuero** [reply]
Attached is an input file that avoids the minimum time step failure before the ball makes contact with the rubber base. I increased the friction coefficient to get there. Since this is a quasi-static simulation, I wonder if small coefficients of friction will lead to the rubber sliding off contact creating an instability --don't have much experience with this problem. 

As @lindsayad  points out, we do second-order routinely, but mostly with quads and sometimes hexes. I can't get the second-order mesh to do contact properly, so I am creating an issue for it. There is an `assert` throwing in `libmesh`.

[dock_slippy.friction.txt](https://github.com/idaholab/moose/files/5898405/dock_slippy.friction.txt)
![dock friction](https://user-images.githubusercontent.com/14300313/106363787-e3939700-62e7-11eb-8259-978581d28ce6.gif)

**Reply by makeclean** [reply]
Thanks :) Any chance you could drop some (presumably) Cubit/Trelis hints on hex meshing for this particular problem?

**Reply by recuero** [reply]
Not really an expert in meshing, but there are some journal files for hexed spheres available on github that could be used for guidance.

**Reply by aeslaughter** [reply]
We need to add this to the  gallery!

**Reply by makeclean** [reply]
@recuero I got a hex version of my problem meshed and I see the exact same behaviour for 1st/2nd order meshes, i.e. for 1st the calculation runs for a bit, makes contact and then eventually takes tiny time steps and fails with the 2nd order one failing when contact is made.

**Reply by recuero** [reply]
@makeclean Any chance you can share the hex mesh/es? I have run relatively complex first order problems with hex meshes without any issues. In principle, I'd think the first order mesh should work. For the second-order case, it does look like nasty debugging is in order. If I get the mesh and input deck, I can see if it stops at the same place in the code. Thanks for trying this.

**Reply by makeclean** [reply]
Could you send me your email address? I should point out for the behaviour above, I was using my original input, not your modified one.

**Reply by recuero** [reply]
You can use this address: amrecuero85 at gmail dot com.

**Reply by GregVernon** [reply]
@makeclean If you're wondering how to mesh a sphere with hexes in Cubit, there's a `sphere` mesh scheme that has the logical structure of 7 cubes (one in the middle, one for each face) as shown below:

![image](https://user-images.githubusercontent.com/868479/107125188-93778000-6865-11eb-98c1-83d54fcad7b5.png)

`volume <vol_id> scheme sphere` 

Will produce the mesh shown on the right:
![image](https://user-images.githubusercontent.com/868479/107125228-de919300-6865-11eb-8e59-879ca6fcaaa6.png)

**Comment by makeclean** [comment]
No, that problem runs fine. I cant'reproduce the contact memory  usage
problem, except in the original geometry.

**Comment by bwspenc** [comment]
I'm jumping into this pretty late. It sounds like some of the memory issues have been resolved. I'd just add here that the `patch_size` parameter in the Mesh block can affect memory usage. That controls the size of the set of nodes on the primary face nearest to the secondary node that contains the faces that are considered for contact with that node. Back in the day, you had to set this really large when you had large sliding between the surfaces, because that set of nodes was static, but now that we have the `patch_update_strategy=iteration` option, that is dynamically updated, so you can get by with a pretty small patch. If I remember right, the set of candidate faces consists of all of the faces that contain the nodes that are found in the patch. That means that you can probably set it to something close to the number of nodes on a face. Maybe you can get by with a patch size of 10 or less.
